### DISCUSSION TITLE
[SSR] NextJS seems to run functions even when not used

### DISCUSSION THREAD

**Discussion by TristanPouliquen** [vercel/next.js#45605] [Help] [answered]
Hello,

I am trying to build a [Tauri](https://tauri.app/) app, with NextJS as the frontend framework. 

Tauri works by bundling a Rust backend, and providing some functions to interact with it from the JS frontend, mainly an `invoke` function that returns a Promise.

These functions are only injected when in the context of a Tauri window, so not working in the SSR mode.

I have created hooks to group my calls to my backend by models, grouping them in a custom hook as you can see here: https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/_state/positions.ts

When I started doing that, I started getting an error stating "Reference error: window is undefined", pointing to a minimized chunk of `@tauri-apps/api` package. 

At first I thought I was maybe calling these methods in a way that did not ensure they are only called dynamically (through a `useEffect` as [here](https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/pages/settings.tsx#L9) or a function called on an event as [here](https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/components/color_picker.tsx#L14)).

So, I decided to try & remove the code calling this custom hook. I still got the same `Reference error: window is undefined`.

DO any of you have an idea of what could cause this erorr? Why does NextJS SSR seem to execute methods from a custom hook that are not even called?

Thank you in advance !

**Comment by icyJoseph** [comment, marked as answer]
I wouldn't make any assumptions at this point. 

Are you able to pin-point the line that's trying to read from window? Any surrounding code? 

I wonder if during bundling perhaps there's a module that does something like:

```js
const foo = window.foo

export const noop = () => {}
```

And then when you import `noop` on a server environment, the module crashes because there's a top level declaration that tries to use window. That's just how modules work, and the way to fix it is to make sure imports cannot possibly collide like this. 

Often the problem is introduced by trying to bring in a type, like:


```ts
const foo = window.foo

export type Bar = string
export const noop = () => {}
```

Then a server side module wants Bar and does:

```ts
import {Bar} from './foo'
```

And it fails. This should take care of it:

```ts
import type {Bar} from './foo'
```

I am myself making a bunch of assumptions here, and the above breakages can sometimes be fixed by the bundler, but yeah, do try to figure out if a server side bundle is importing client code with top level window access.

## Edit

OP had to change the importing statements as expected. Instead of `@tauri-apps/api` OP had to import `invoke` from `@tauri-apps/api/tauri` and that skips importing the `window` module from Tauri, which does some top level raw check on `window`.

**Reply by TristanPouliquen** [reply]
The minified code where the error happens is:
```
"__TAURI_METADATA__"in window?b=new s(window.__TAURI_METADATA__.__currentWindow.label,{skip:!0}):(console.warn(`Could not find "window.__TAURI_METADATA__". The "appWindow" value will reference the "main" window label.
Note that this is not an issue if running this frontend on a browser instead of a Tauri window.`),b=new s("main",{skip:!0}));
```

which comes from: https://github.com/tauri-apps/tauri/blob/dev/tooling/api/src/window.ts#L2033

But why would NextJS load this in the SSR mode when it is not called at all in my code?

**Reply by icyJoseph** [reply]
It could still be imported by a tauri module, that's the fun thing with modules, you know never what comes in when you import them.

In a subtle way the issue is that the module you linked does a raw check:

```js
if ('__TAURI_METADATA__' in window) {
  appWindow = new WebviewWindow(
    window.__TAURI_METADATA__.__currentWindow.label,
    {
      // @ts-expect-error `skip` is not defined in the public API but it is handled by the constructor
      skip: true
    }
  )
} 
```

Without checking if `window` is available in the first place.

You could, tell Webpack to ignore, the path in SSR, but I am not sure how does the actual import path look like:

```js
  webpack(config, {nextRuntime}) {
        if (typeof nextRuntime === 'nodejs') { // maybe edge should be considered too
            config.resolve.fallback = {
                ...config.resolve.fallback,
                async_hooks: false, // ignore node async_hooks, which is a node API
            };
            // manually ignore some local modules and node_modules
            const {IgnorePlugin} = require('webpack');
            // the module that imports window 
            const ignoreUserModule = new IgnorePlugin({resourceRegExp: /some\/user\/module/});
       
            config.plugins.push(ignoreUserModule);
        }

        return config;
    },
```

Is it `@tauri-apps/api` that is importing the `window.ts` module? Please understand that this is not Next.js running your code, but rather that importing a module, executes its top level expressions, in this case: `if ('__TAURI_METADATA__' in window)` is at the top level, and that's the reason it trips. ~~ I am a bit busy today so I may be posting inaccurate statements, just so you know 😅

**Reply by TristanPouliquen** [reply]
Thank you for your time and explations! I do not know all the intrincate parts of module loading indeed :-) 

0ne of the maintainers of Tauri answered me, and this code should only be executed when loading the `window` module, which I do not.

This started happening when I installed RecoilJs for my state management & reordered my files differently. 

He thinks the issue might be that Recoil tries to mount the component while in SSR mode, and that this might lead to bad behaviour. 

I will keep pushing to see what I can find.

**Reply by icyJoseph** [reply]
But in this entry point file, https://github.com/tauri-apps/tauri/blob/dev/tooling/api/src/index.ts, I see:

```js
import * as app from './app'
import * as cli from './cli'
import * as clipboard from './clipboard'
import * as dialog from './dialog'
import * as event from './event'
import * as fs from './fs'
import * as globalShortcut from './globalShortcut'
import * as http from './http'
import * as notification from './notification'
import * as path from './path'
import * as process from './process'
import * as shell from './shell'
import * as tauri from './tauri'
import * as updater from './updater'
import * as window from './window'
import * as os from './os'

/** @ignore */
const invoke = tauri.invoke

export {
  invoke,
```

So when you import:

```js
import { invoke } from "@tauri-apps/api";
```

In turn you import `* as window` and it breaks, because that module has a top level "raw" window check. 

Could you try with:

```js
import { invoke } from "@tauri-apps/api/invoke";
```

**Reply by TristanPouliquen** [reply]
I get an error stating that this file is not exported :
`Module not found: Package path ./invoke is not exported from package /home/tristan/dev/perso/selection-poste-enm/node_modules/@tauri-apps/api (see exports field in /home/tristan/dev/perso/selection-poste-enm/node_modules/@tauri-apps/api/package.json)
`

I'm beginning to think that I don't really need heavy SSR optimization like NextJS can propose, so I might just switch to a basic React app. I'll have 3/4 pages tops, with not so much data to load.

Thank you for your time nonetheless!

**Reply by TristanPouliquen** [reply]
I think we finally got it, I have to import the `invoke` method from `@tauri-apps/api/tauri`, and this does not load the `window` that was causing the issue. 

Thank you again for the time you took to help me!

**Reply by icyJoseph** [reply]
Awesome! I'll update my answer then, please don't forget to mark as answered! Cool app btw, looks very interesting!

**Reply by TristanPouliquen** [reply]
Done! 

Yes, I had already worked with Electron, but wanted to give it a go, to test out Rust & see the difference in terms of memory & space usage

**Reply by icyJoseph** [reply]
Yeah I once tested out a CRA on Tauri, and it was cool, but I didn't really see any major wins, the app was kind of small anyway. 

Are you planning to pen an article? Kind of curious about your findings.

**Reply by TristanPouliquen** [reply]
Not sure, my app is really small too, and for private use, so I was just using it to toy with the latests techs & frameworks, I won't be looking so much at performance, etc
